### ISSUE TITLE
ZIMs without media (text-based) throw an error on attempting to read the MIME type list

### ISSUE THREAD

**Issue by Jaifroid** [kiwix/kiwix-js#628]
On attempting to open a text-based ZIM such as ~~`wikipedia_en_medicine_maxi_2020-06.zim`~~ `wikipedia_es_all_mini_2020-06.zim`, I get the following error in `zimFile.js`:

![image](https://user-images.githubusercontent.com/4304337/85918174-3db31200-b858-11ea-996b-e394ddbcf454.png)

I guess we need explicitly to handle cases where there is no MIME type list because there are no media in the ZIM.

**Comment by Jaifroid** [comment]
PS The error does not block using the app with the ZIM, but we should still handle it.

**Comment by kelson42** [comment]
I hardly see how this could be a valid ZIM file. You talk about https://wiki.openzim.org/wiki/ZIM_file_format#MIME_Type_List_.28mimeListPos.29?

**Comment by Jaifroid** [comment]
@kelson42 Yes, that is the list the app reads into memory on first loading the ZIM file. Possibly the list is zero-length, and we simply don't handle it properly. The ZIM in question, `wikipedia_en_medicine_maxi_2020-06.zim`, is still perfectly readable in Kiwix JS.

**Comment by Jaifroid** [comment]
@kelson42 Apologies, I mis-copied the ZIM. It's `wikipedia_es_all_mini_2020-06.zim`.

**Comment by kelson42** [comment]
@Jaifroid HTML pages having a mime-type as well, I don't understand how legitimly this list could be empty. The only use case would be a totally empty ZIM file (which is an edge case which I believe is not fully supported properly - but this is an other story).

**Comment by Jaifroid** [comment]
Most likely an issue with the function we use to read the list. I'll debug and get back. I guess it *could* be a malformed ZIM file, but I wouldn't jump to that conclusion before debugging our custom function.

**Comment by Jaifroid** [comment]
After a lot of debugging, this issue appears in fact to be caused by the same thing that is causing the freeze on opening recent ZIM files in Kiwix JS (#621), and it is also causing (I believe) the issue reported in #629. **The problem appears to be that our code assumes that the URL Pointer List comes after the MIME Type List in the ZIM file.** This was always the case, but in the most recent files the URL Pointer List appears to have been relocated to the very end of the ZIM file. This explains why our readers are failing on these new ZIMs (especially if they are short on memory). They get the size of the MIME Type List by assuming that the URL Pointer List comes immediately after the header. This assumption is incorrect in the new ZIMs being produced in June (and some produced earlier). 

Take the example of `wikipedia_es_all_mini_2020-06`. This file has a file size of 2,753,296,905 bytes. According to 8 bytes at offset 32 of the file header, the URL Pointer List should be found at position 2,714,681,137, i.e. right at the end of the file. In previous ZIM files, the URL Pointer List would start straight after the header, so it was a safe bet that to find out how many bytes to read to get the full MIME Type list, we should get a Uint8Array starting at position 80 (after the header) and finishing at UrlPtrPos. **This assumption is wrong!**. It is causing Kiwix JS to read from position 80 all the way through the file to position 2,714,681,137, which being around 2GB could easily fail depending on the specification of the system.

**NB LibZim makes the same wrong assumption: see https://github.com/openzim/libzim/blob/master/src/fileimpl.cpp#L275**

@kelson42 @Popolechien I think this explains the issue with the recent ZIMs including the issue with the Spanish ZIMs we discussed elsewhere.

We need instead to read a conservative number of bytes that will always be larger than any imaginable MIME Type list, but not so large as to cause memory issues. When I first wrote this function, I chose 256 bytes, but @mossroy [asked me to follow LibZim](https://github.com/kiwix/kiwix-js/pull/490#discussion_r280993945), which was good advice.

But it looks like we may have to revert to a hard-coded value.

**Comment by Jaifroid** [comment]
@kelson42 Based on my limited understanding of C++, I think the same issue is in the libZim codebase. Please take a look at:
https://github.com/openzim/libzim/blob/master/src/fileimpl.cpp#L130, and in particular these lines:
```
    // read mime types
    size = zsize_t(header.getUrlPtrPos() - header.getMimeListPos());
    // No need to check access, getUrlPtrPos is in the zim file, and we are
    // sure that getMimeListPos is 80.
    auto buffer = zimReader->get_buffer(offset_t(header.getMimeListPos()), size);
    offset_t current = offset_t(0);
```
It looks very much to me like the size of the MIME type list is being calculated by subtracting the `mimeListPos` from `urlPtrPos`. This works fine **only** if the URL pointer list starts one byte after the end of the MIME type list. However, if the URL pointer list is relocated elsewhere, e.g. at the end of the ZIM file, then this will become a huge size and the reader may well fail to load the MIME type list depending on size of ZIM file and on memory in the machine. Several recent ZIMs have the URL pointer list at the end of the file,

I have now added [a patch](https://github.com/kiwix/kiwix-js-windows/commit/2406b1c4f62d50ee8bc341d92126d310d4366d21) to Kiwix JS Windows to test this. It can be tested by following instructions here https://github.com/kiwix/kiwix-js/issues/629#issuecomment-650705479. Kiwix JS Windows is now able to read mwoffliner ZIMs coded in June and the patch completely fixes the freeze reported in #621.

**Comment by Jaifroid** [comment]
@kelson42 Thanks for letting me know!
